// Electricity/src/elecData/Vertex.ts
/**
 * Created by hanyeah on 2019/8/30.
 */
namespace hanyeah.electricity.elecData {
  import ListNode = hanyeah.dataStruct.ListNode;
  import UnionFindSet = hanyeah.dataStruct.UnionFindSet;
  export class Vertex extends HObject {
    public U: number = 0;
    public connLN: ListNode;
    public connUFS: UnionFindSet;
    public worldLN: ListNode;
    public graphUFS: UnionFindSet;
    public world: World;
    constructor(world: World) {
      super();
      this.connLN = new ListNode(this);
      this.connUFS = new UnionFindSet(this);
      this.worldLN = new ListNode(this);
      this.graphUFS = new UnionFindSet(this);
      this.world = world;
      world.addVertex(this);
    }

    destroy() {
      this.world.removeVertex(this);
      this.world = null;
      this.disConnect();
      this.connLN.destroy();
      this.connLN = null;
      this.connUFS.destroy();
      this.connUFS = null;
      this.graphUFS.destroy();
      this.graphUFS = null;
      this.worldLN.destroy();
      this.worldLN = null;
    }

    connect(vertex: Vertex): void {
      if (this.connUFS.root !== vertex.connUFS.root) {
        this.connLN.connect(vertex.connLN);
        this.connUFS.root.root = vertex.connUFS.root;
      }
    }

    disConnect(): void {
      if (this.connUFS.isRoot()) {
        let next: ListNode = this.connLN.next;
        let newRoot: UnionFindSet = (next.userData as Vertex).connUFS;
        while (next !== this.connLN) {
          (next.userData as Vertex).connUFS.root = newRoot;
          next = next.next;
        }
      }
      this.connLN.disConnect();
    }
  }
}
