// Electricity/src/dataStruct/List.ts
/**
 * Created by hanyeah on 2019/8/30.
 */
namespace hanyeah.dataStruct {
  export class List {
    private guard: ListNode;
    private changed: boolean = false;
    private nodes: ListNode[] = [];
    private userDatas: object[] = [];
    constructor() {
      this.guard = new ListNode(null);
    }

    public destroy() {
      this.guard = null;
      this.nodes = null;
      this.userDatas = null;
    }

    public add(node: ListNode): void {
      this.changed = true;
      this.guard.connect(node);
    }

    public remove(node: ListNode): void {
      this.changed = true;
      node.disConnect();
    }

    public getAll(): ListNode[] {
      this.updateAry();
      return this.nodes;
    }

    public getAllUserData(): object[] {
      this.updateAry();
      return this.userDatas;
    }

    private updateAry(): void {
      if (this.changed) {
        this.changed = false;
        this.nodes = [];
        this.userDatas = [];
        let node: ListNode = this.guard.prev;
        while (node !== this.guard){
          this.nodes.push(node);
          this.userDatas.push(node.userData);
          node = node.prev;
        }
      }
    }

  }
}
